File: gen_blockstate_jsons.pl
#!/usr/bin/perl

=head1 NAME
    
gen_blockstate_jsons.pl - create a blockstate json.

=head1 SYNOPSIS

gen_blockstate_jsons.pl modid [topdir]

=over

=item modid

mandatory modid tag; to be prefixed to various entries.

=item topdir

Optional root directory for output jsons-usually the assets/modid/ 
directory. Uses current working directory if not specified.

=back

=head1 DESCRIPTION

Create a blockstate json for B<modid> by following a set of prompts
for name, type, etc.  Split out from the old build_graphic_jsons.pl
script and updated.

=cut

use strict;
use warnings;

use Cwd;
use File::Spec;
use File::Copy;
use File::Path qw(make_path);
use List::MoreUtils qw(uniq);

use FindBin;
use lib "$FindBin::Bin";
use generator;

# output directory is...
my $TOP_LEVEL = defined($ARGV[1]) ? $ARGV[1] : getcwd();
my $BLOCKSTATE_PATH = "${TOP_LEVEL}/blockstates";
my $BLOCK_MODEL_PATH = "${TOP_LEVEL}/models/block";
my $ITEM_MODEL_PATH = "${TOP_LEVEL}/models/item";

my $MODID = $ARGV[0];

my $verbose = 1;

die "modid is required!" if ! defined($MODID);

# create if they do not exist.
make_path($BLOCKSTATE_PATH);

# go there
chdir $TOP_LEVEL or die "Unable to change to ${TOP_LEVEL}: $!\n";

my $resp;
my $block_subtype = 0;
my $bs_template;
my $block_template;
my $not_done = 1;

while ($not_done)
{
    print "Generic simple block (G;default), ",
        "crop block [has growth stages] (C), ",
        "Simple block with facing [like oak_log] (F), ",
        "Bars [like iron_bars] (B), Doors (D), ", 
        "Thin pane [like glass] (T), Stairs (S), Other (O) ? ";
    $resp = <STDIN>;
    $block_subtype = ($resp =~ /^[GCFBTSDO]/i) ? uc(substr($resp,0,1)) : 'G';

    if ($verbose) {
        print "\$block_subtype = ", $block_subtype, "\n";
    }

    get_block_info($block_subtype);
    $not_done = check_repeat("Create another blockstate [Y/N]? ");
} ## end-while not_done

exit;

=head1 FUNCTIONS

=over

=item B<get_block_info>

Actually does most of the work of this script. 
=cut

sub get_block_info
{
    my $block_type = $_[0];
    my $prompt;
    my $templ_model_stem;
    my $not_done = 1;
    
    if ($block_type eq 'G')  # simple generic cube block
    {
        my @out_jsons;
        while ($not_done)
        {
            $prompt = "Name of block to create blockstate for (e.g. foo_block): ";
            @out_jsons = get_simple_json_path($prompt, $BLOCKSTATE_PATH);
            write_cubeall_blockstate($MODID, @out_jsons);
            $not_done = check_repeat("Create another simple cube [Y/N]? ");
        } ## end while not_done 
    } ## end-if simple cube
    elsif ($block_type eq 'F')  # simple block with horizontal facing
    {
        make_blockstate_prompts('oak_log');
    } ## end-elsif 'F'            
    elsif ($block_type eq 'C') # crop block (has growth stages).
    {
        make_blockstate_prompts('carrots');
    }
    elsif ($block_type eq 'B') # metal bars
    {
        make_blockstate_prompts('iron_bars');
    }
    elsif ($block_type eq 'S') # stairs
    {
        make_blockstate_prompts('oak_stairs');
    }
    elsif ($block_type eq 'D') #doors
    {
        make_blockstate_prompts('oak_door');
    }
    elsif ($block_type eq 'T') #thin pane
    {
        make_blockstate_prompts('glass_pane');
    }
    else # not a simple anything 
    {
        $prompt = "Vanilla block to use as template (e.g. acacia_fence): ";
        my $block_stem = get_response($prompt);
        make_blockstate_prompts($block_stem);
    } ## end-else not a simple cube
} ## end sub get_block_info


=item B<make_blockstate_prompts>

Bundles up a bunch of repetitive, common code for making blockstates.

=cut

sub make_blockstate_prompts
{
    my ($block_stem, ) = @_;

    my $not_done = 1;
    my $template_json = get_template($block_stem, $MC_BLOCKSTATE_PATH);
    die if ! defined $template_json;
    print("Source blockstate: ", $template_json, "\n") if $verbose;
    while ($not_done)
    {
        my $prompt = "Name of block to create files for (e.g. foo_gourd): ";
        my @out_jsons = get_simple_json_path($prompt, $BLOCKSTATE_PATH);
        print("Target blockstate: ", $out_jsons[1], "\n") if $verbose;
        my $out_model_stem = $out_jsons[0];
        printf("Target block stem to replace %s with: %s\n", 
                $block_stem, $out_model_stem) if $verbose;

        # write blockstate file
        copy_blockstate($template_json, $block_stem, $out_jsons[1],
                        $out_model_stem);
        # repeat?
        $not_done = check_repeat(
                    "Create another block from the same template [Y/N]? ");
    } ## end while not_done 

} ## end sub make_blockstate_prompts


=item B<write_cubeall_blockstate>

write a standard Forge blockstate for a stock model block.

=cut

sub write_cubeall_blockstate
{
    my ($modid, $name, $out_json) = @_;
    open (my $fh, ">", $out_json) or die "Unable to open ${out_json}: $!";
    print $fh "{\n";
    print $fh "\t\"variants\": {\n";
    print $fh "\t\t\"\": { \"model\": \"${modid}:block/${name}\" }\n";
    print $fh "\t}\n";
    print $fh "}\n";
    close $fh;
} ## end write_cubeall_blockstate()


=item B<copy_blockstate>

Copies from one blockstate to another, substituting the block name
stem C<$out_mstem> for C<$in_mstem>, and prefixing the modid.

Parameters:
    
    $in_json - full path to source blockstate json.
    $in_mstem - block name stem to be replaced.
    $out_json - full path to target blockstate json.
    out_mstem - block name stem to replace $in_mstem with.

=cut

sub copy_blockstate
{
    my ($in_json, $in_mstem, $out_json, $out_mstem ) = @_;
    open (my $fh, "<", $in_json) or die "Unable to open ${in_json}: $!";
    open (my $fh2, ">", $out_json) or die "Unable to open ${out_json}: $!";
    while (my $line = <$fh>)
    {
        $line =~ s/block\/${in_mstem}/${MODID}:block\/${out_mstem}/;
        print $fh2 $line;
    }
    close $fh;
    close $fh2;
} ## end copy_blockstate()


=back

=head1 HISTORY

=cut

File: gen_blockstate_jsons.py
#!/usr/bin/python3
"""
Generate blockstate jsons. Replaces old gen_blockstate_jsons.pl perl script.\

Must be run in src/main/resources/assets/<modid>/blockstates.

SYNOPSIS:

gen_blockstate_jsons.py [-h] --type
                               {simple,crop,facing,bars,doors,pane,stairs,pressure_plate,other}
                               blockname

Generate blockstates for standard block types

positional arguments:
  blockname             blockstate filename

optional arguments:
  -h, --help            show this help message and exit
  --type {simple,crop,facing,bars,doors,pane,stairs,pressure_plate,other}, -t {simple,crop,facing,bars,doors,pane,stairs,pressure_plate,other}
                        type of blockstate

"""

import sys
import os
import os.path
import argparse
import json
import copy

# TEMPLATES

BLOCK_TEMPLATE = { 'variants' : { '' : { "model" : None } } }

PRESSURE_PLATE = { 'variants' : { 
    'powered=false' : { "model" : None },
    'powered=true' : { "model" : None } 
} }

SLAB_TEMPLATE = { 'variants' : { 
    'type=bottom' : { "model" : None },
    'type=double' : { "model" : None },
    'type=top' : { "model" : None },
} }

BARS_TEMPLATE = {   "multipart": [
    { "apply": { "model": None } },
    { "when": { "north": "false", "west": "false", "south": "false", "east": "false" },
      "apply": { "model": None }
    },
    { "when": { "north": "true", "west": "false", "south": "false", "east": "false" },
      "apply": { "model": None }
    },
    { "when": { "north": "false", "west": "false", "south": "false", "east": "true" },
      "apply": { "model": None, "y": 90 }
    },
    { "when": { "north": "false", "west": "false", "south": "true", "east": "false" },
      "apply": { "model": None }
    },
    { "when": { "north": "false", "west": "true", "south": "false", "east": "false" },
      "apply": { "model": None, "y": 90 }
    },
    { "when": { "north": "true" }, 
      "apply": { "model": None } },
    { "when": { "east": "true" },
      "apply": { "model": None, "y": 90 }
    },
    { "when": { "south": "true" },
      "apply": { "model": None }
    },
    { "when": { "west": "true" },
      "apply": { "model": None, "y": 90 }
    }
] }

STAIRS_TEMPLATE = { "variants": {
    "facing=east,half=bottom,shape=straight":  { "model": None },
    "facing=west,half=bottom,shape=straight":  { "model": None, "y": 180, "uvlock": True },
    "facing=south,half=bottom,shape=straight": { "model": None, "y": 90, "uvlock": True },
    "facing=north,half=bottom,shape=straight": { "model": None, "y": 270, "uvlock": True },
    "facing=east,half=bottom,shape=outer_right":  { "model": None },
    "facing=west,half=bottom,shape=outer_right":  { "model": None, "y": 180, "uvlock": True },
    "facing=south,half=bottom,shape=outer_right": { "model": None, "y": 90, "uvlock": True },
    "facing=north,half=bottom,shape=outer_right": { "model": None, "y": 270, "uvlock": True },
    "facing=east,half=bottom,shape=outer_left":  { "model": None, "y": 270, "uvlock": True },
    "facing=west,half=bottom,shape=outer_left":  { "model": None, "y": 90, "uvlock": True },
    "facing=south,half=bottom,shape=outer_left": { "model": None },
    "facing=north,half=bottom,shape=outer_left": { "model": None, "y": 180, "uvlock": True },
    "facing=east,half=bottom,shape=inner_right":  { "model": None },
    "facing=west,half=bottom,shape=inner_right":  { "model": None, "y": 180, "uvlock": True },
    "facing=south,half=bottom,shape=inner_right": { "model": None, "y": 90, "uvlock": True },
    "facing=north,half=bottom,shape=inner_right": { "model": None, "y": 270, "uvlock": True },
    "facing=east,half=bottom,shape=inner_left":  { "model": None, "y": 270, "uvlock": True },
    "facing=west,half=bottom,shape=inner_left":  { "model": None, "y": 90, "uvlock": True },
    "facing=south,half=bottom,shape=inner_left": { "model": None },
    "facing=north,half=bottom,shape=inner_left": { "model": None, "y": 180, "uvlock": True },
    "facing=east,half=top,shape=straight":  { "model": None, "x": 180, "uvlock": True },
    "facing=west,half=top,shape=straight":  { "model": None, "x": 180, "y": 180, "uvlock": True },
    "facing=south,half=top,shape=straight": { "model": None, "x": 180, "y": 90, "uvlock": True },
    "facing=north,half=top,shape=straight": { "model": None, "x": 180, "y": 270, "uvlock": True },
    "facing=east,half=top,shape=outer_right":  { "model": None, "x": 180, "y": 90, "uvlock": True },
    "facing=west,half=top,shape=outer_right":  { "model": None, "x": 180, "y": 270, "uvlock": True },
    "facing=south,half=top,shape=outer_right": { "model": None, "x": 180, "y": 180, "uvlock": True },
    "facing=north,half=top,shape=outer_right": { "model": None, "x": 180, "uvlock": True },
    "facing=east,half=top,shape=outer_left":  { "model": None, "x": 180, "uvlock": True },
    "facing=west,half=top,shape=outer_left":  { "model": None, "x": 180, "y": 180, "uvlock": True },
    "facing=south,half=top,shape=outer_left": { "model": None, "x": 180, "y": 90, "uvlock": True },
    "facing=north,half=top,shape=outer_left": { "model": None, "x": 180, "y": 270, "uvlock": True },
    "facing=east,half=top,shape=inner_right":  { "model": None, "x": 180, "y": 90, "uvlock": True },
    "facing=west,half=top,shape=inner_right":  { "model": None, "x": 180, "y": 270, "uvlock": True },
    "facing=south,half=top,shape=inner_right": { "model": None, "x": 180, "y": 180, "uvlock": True },
    "facing=north,half=top,shape=inner_right": { "model": None, "x": 180, "uvlock": True },
    "facing=east,half=top,shape=inner_left":  { "model": None, "x": 180, "uvlock": True },
    "facing=west,half=top,shape=inner_left":  { "model": None, "x": 180, "y": 180, "uvlock": True },
    "facing=south,half=top,shape=inner_left": { "model": None, "x": 180, "y": 90, "uvlock": True },
    "facing=north,half=top,shape=inner_left": { "model": None, "x": 180, "y": 270, "uvlock": True }
} }

DOORS_TEMPLATE = {  "variants": {
    "facing=east,half=lower,hinge=left,open=false": { "model": None },
    "facing=east,half=lower,hinge=left,open=true": { "model": None, "y": 90 },
    "facing=east,half=lower,hinge=right,open=false": { "model": None },
    "facing=east,half=lower,hinge=right,open=true": { "model": None, "y": 270 },
    "facing=east,half=upper,hinge=left,open=false": { "model": None },
    "facing=east,half=upper,hinge=left,open=true": { "model": None, "y": 90 },
    "facing=east,half=upper,hinge=right,open=false": { "model": None },
    "facing=east,half=upper,hinge=right,open=true": { "model": None, "y": 270 },
    "facing=north,half=lower,hinge=left,open=false": { "model": None, "y": 270 },
    "facing=north,half=lower,hinge=left,open=true": { "model": None },
    "facing=north,half=lower,hinge=right,open=false": { "model": None, "y": 270 },
    "facing=north,half=lower,hinge=right,open=true": { "model": None, "y": 180 },
    "facing=north,half=upper,hinge=left,open=false": { "model": None, "y": 270 },
    "facing=north,half=upper,hinge=left,open=true": { "model": None },
    "facing=north,half=upper,hinge=right,open=false": { "model": None, "y": 270 },
    "facing=north,half=upper,hinge=right,open=true": { "model": None, "y": 180 },
    "facing=south,half=lower,hinge=left,open=false": { "model": None, "y": 90 },
    "facing=south,half=lower,hinge=left,open=true": { "model": None, "y": 180 },
    "facing=south,half=lower,hinge=right,open=false": { "model": None, "y": 90 },
    "facing=south,half=lower,hinge=right,open=true": { "model": None },
    "facing=south,half=upper,hinge=left,open=false": { "model": None, "y": 90 },
    "facing=south,half=upper,hinge=left,open=true": { "model": None, "y": 180 },
    "facing=south,half=upper,hinge=right,open=false": { "model": None, "y": 90 },
    "facing=south,half=upper,hinge=right,open=true": { "model": None },
    "facing=west,half=lower,hinge=left,open=false": {
      "model": None,
      "y": 180
    },
    "facing=west,half=lower,hinge=left,open=true": {
      "model": None,
      "y": 270
    },
    "facing=west,half=lower,hinge=right,open=false": {
      "model": None,
      "y": 180
    },
    "facing=west,half=lower,hinge=right,open=true": {
      "model": None,
      "y": 90
    },
    "facing=west,half=upper,hinge=left,open=false": {
      "model": None,
      "y": 180
    },
    "facing=west,half=upper,hinge=left,open=true": {
      "model": None,
      "y": 270
    },
    "facing=west,half=upper,hinge=right,open=false": {
      "model": None,
      "y": 180
    },
    "facing=west,half=upper,hinge=right,open=true": {
      "model": None,
      "y": 90
    }
}}

# command-line arguments
parser = argparse.ArgumentParser(description="Generate blockstates for standard block types")
parser.add_argument("blockname", help="blockstate filename")
parser.add_argument("--type", "-t", 
        choices=['simple', 'crop', 'facing', 'bars', 'door', 'pane', 'stairs',
                 'pressure_plate', 'slab', 'other'], 
        help="type of blockstate", required=True)

args = parser.parse_args()
#print(args);

# parse directory and make sure we are in the right place...
(head, tail) = os.path.split(os.getcwd())
if tail != 'blockstates':
    print('Warning: not in blockstates directory')
    exit()
(head, modid) = os.path.split(head)
(head, tail) = os.path.split(head)
if tail != 'assets':
    print('Warning: not in assets/{}/blockstates directory'.format(modid))
    exit()

# construct the blockstate

if args.type == 'simple':
    blockstate = copy.deepcopy(BLOCK_TEMPLATE)
    blockstate['variants']['']['model'] = "{}:block/{}".format(modid, args.blockname)
elif args.type == 'pressure_plate':
    blockstate = copy.deepcopy(PRESSURE_PLATE)
    blockstate['variants']['powered=false']['model'] = "{}:block/{}".format(modid, args.blockname)
    blockstate['variants']['powered=true']['model'] = "{}:block/{}_down".format(modid, args.blockname)
elif args.type == 'slab':
    blockstate = copy.deepcopy(SLAB_TEMPLATE)
    lblockname = args.blockname
    if lblockname.endswith('_block'):
        nn = lblockname.rindex('_block');
        lblockname = lblockname[0:nn]
    blockstate['variants']['type=bottom']['model'] = "{}:block/{}".format(modid, "{}_slab".format(lblockname))
    blockstate['variants']['type=double']['model'] = "{}:block/{}".format(modid, args.blockname)
    blockstate['variants']['type=top']['model'] = "{}:block/{}".format(modid, "{}_slab_top".format(lblockname))
elif args.type == 'bars':
    blockstate = copy.deepcopy(BARS_TEMPLATE)
    post_ends = "{}:block/{}_post_ends".format(modid, args.blockname)
    post = "{}:block/{}_post".format(modid, args.blockname)
    cap = "{}:block/{}_cap".format(modid, args.blockname)
    cap_alt = "{}:block/{}_cap_alt".format(modid, args.blockname)
    side = "{}:block/{}_side".format(modid, args.blockname)
    side_alt = "{}:block/{}_side_alt".format(modid, args.blockname)
    blockstate['multipart'][0]['apply']['model'] = post_ends
    blockstate['multipart'][1]['apply']['model'] = post
    blockstate['multipart'][2]['apply']['model'] = cap
    blockstate['multipart'][3]['apply']['model'] = cap
    blockstate['multipart'][4]['apply']['model'] = cap_alt
    blockstate['multipart'][5]['apply']['model'] = cap_alt
    blockstate['multipart'][6]['apply']['model'] = side
    blockstate['multipart'][7]['apply']['model'] = side
    blockstate['multipart'][8]['apply']['model'] = side_alt
    blockstate['multipart'][9]['apply']['model'] = side_alt

elif args.type == 'stairs':
    blockstate = copy.deepcopy(STAIRS_TEMPLATE)
    lblockname = args.blockname
    if lblockname.endswith('_stairs'):
        nn = lblockname.rindex('_stairs');
        lblockname = lblockname[0:nn]
    stairs = "{}:block/{}_stairs".format(modid, lblockname)
    inner_stairs = "{}:block/{}_inner_stairs".format(modid, lblockname)
    outer_stairs = "{}:block/{}_outer_stairs".format(modid, lblockname)
    blockstate['variants']["facing=east,half=bottom,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=west,half=bottom,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=south,half=bottom,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=north,half=bottom,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=east,half=bottom,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=west,half=bottom,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=south,half=bottom,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=north,half=bottom,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=east,half=bottom,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=west,half=bottom,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=south,half=bottom,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=north,half=bottom,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=east,half=bottom,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=west,half=bottom,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=south,half=bottom,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=north,half=bottom,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=east,half=bottom,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=west,half=bottom,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=south,half=bottom,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=north,half=bottom,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=east,half=top,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=west,half=top,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=south,half=top,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=north,half=top,shape=straight"]['model'] = stairs
    blockstate['variants']["facing=east,half=top,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=west,half=top,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=south,half=top,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=north,half=top,shape=outer_right"]['model'] = outer_stairs
    blockstate['variants']["facing=east,half=top,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=west,half=top,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=south,half=top,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=north,half=top,shape=outer_left"]['model'] = outer_stairs
    blockstate['variants']["facing=east,half=top,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=west,half=top,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=south,half=top,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=north,half=top,shape=inner_right"]['model'] = inner_stairs
    blockstate['variants']["facing=east,half=top,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=west,half=top,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=south,half=top,shape=inner_left"]['model'] = inner_stairs
    blockstate['variants']["facing=north,half=top,shape=inner_left"]['model'] = inner_stairs
    
else:
    print("Error: {} not yet implemented\n".format(args.type), file=sys.stderr)
    sys.exit()


#TODO

filename = "{}.json".format(args.blockname)
with open(filename, 'w') as f:
    json.dump(blockstate, f, indent=4, sort_keys=False)

